pkg/cephmgr/rgw/agent.go

```diff
@@ -0,0 +1,258 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package rgw
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"regexp"
+
+	etcd "github.com/coreos/etcd/client"
+	ctx "golang.org/x/net/context"
+
+	"strconv"
+
+	"github.com/rook/rook/pkg/cephmgr/client"
+	"github.com/rook/rook/pkg/cephmgr/mon"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/util"
+	"github.com/rook/rook/pkg/util/proc"
+)
+
+const (
+	rgwPort         = 8123
+	rgwAgentName    = "rgw"
+	keyringTemplate = `[client.radosgw.gateway]
+	key = %s
+	caps mon = "allow rw"
+	caps osd = "allow *"
+`
+)
+
+type rgwAgent struct {
+	factory client.ConnectionFactory
+	rgwProc *proc.MonitoredProc
+}
+
+func NewAgent(factory client.ConnectionFactory) *rgwAgent {
+	return &rgwAgent{factory: factory}
+}
+
+func (a *rgwAgent) Name() string {
+	return rgwAgentName
+}
+
+// set the desired state in etcd
+func (a *rgwAgent) Initialize(context *clusterd.Context) error {
+	return nil
+}
+
+// configure RGW on the local node if it is in the desired state
+func (a *rgwAgent) ConfigureLocalService(context *clusterd.Context) error {
+	// check if the rgw is in desired state on this node
+	desired, err := getRGWState(context.EtcdClient, context.NodeID, false)
+	if (err != nil && util.IsEtcdKeyNotFound(err)) || (err == nil && !desired) {
+		// the rgw is not in the desired state on this node, so ensure it is not running
+		return a.DestroyLocalService(context)
+	} else if err != nil {
+		return fmt.Errorf("failed to load rgw state. %+v", err)
+	}
+
+	// load cluster info
+	cluster, err := mon.LoadClusterInfo(context.EtcdClient)
+	if err != nil {
+		return fmt.Errorf("failed to load cluster info. %+v", err)
+	}
+
+	err = a.generateConfigFiles(context, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to generate config. %+v", err)
+	}
+
+	// start rgw
+	err = a.startRGW(context, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to start rgw process. %+v", err)
+	}
+
+	return nil
+}
+
+func (a *rgwAgent) DestroyLocalService(context *clusterd.Context) error {
+	if a.rgwProc == nil {
+		logger.Debugf("no need to stop rgw that is not running")
+		return nil
+	}
+
+	if err := a.rgwProc.Stop(); err != nil {
+		return fmt.Errorf("failed to stop rgw. %+v", err)
+	}
+
+	logger.Infof("stopped ceph rgw")
+	a.rgwProc = nil
+
+	return nil
+}
+
+func (a *rgwAgent) generateConfigFiles(context *clusterd.Context, cluster *mon.ClusterInfo) error {
+	// create the rgw data directory
+	dataDir := path.Join(getRGWConfDir(context.ConfigDir), "data")
+	if err := os.MkdirAll(dataDir, 0744); err != nil {
+		logger.Warningf("failed to create rgw data directory %s: %+v", dataDir, err)
+	}
+
+	settings := map[string]string{
+		"host":                           "ceph-rgw",
```

Comment:
help me understand how the DNS name is used by an external S3 tool?  `rook object create` is run, which kicks off RGW orchestration.  then `rook object connection` (or something like that) can be run to get all the user creds, endpoint, etc.  Is that enough to put into an external S3 tool, or does /etc/hosts really have to be updated?